pkg/cni/cilium/cilium.go

```diff
@@ -0,0 +1,275 @@
+				Verbs:     []string{"get", "list", "watch", "create"},
+			},
+			{
+				APIGroups: []string{"apiextensions.k8s.io"},
+				Resources: []string{"customresourcedefinitions"},
+				Verbs:     []string{"create", "get", "list", "watch", "update"},
+			},
+			{
+				APIGroups: []string{"cilium.io"},
+				Resources: []string{"ciliumnetworkpolicies", "ciliumnetworkpolicies/status", "ciliumendpoints", "ciliumendpoints/status"},
+				Verbs:     []string{"*"},
+			},
+		},
+	}
+
+	clusterRoleBinding = rbac.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: CiliumClusterRoleName,
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: rbac.GroupName,
+			Kind:     "ClusterRole",
+			Name:     CiliumClusterRoleName,
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind:      rbac.ServiceAccountKind,
+				Name:      CiliumServiceAccountName,
+				Namespace: metav1.NamespaceSystem,
+			},
+			{
+				Kind: "Group",
+				Name: "system:nodes",
+			},
+		},
+	}
+
+	clusterRoleBindingPSP = rbac.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: CiliumClusterRoleNamePSP,
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: rbac.GroupName,
+			Kind:     "ClusterRole",
+			Name:     "suse:kubic:psp:privileged",
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind:      rbac.ServiceAccountKind,
+				Name:      CiliumServiceAccountName,
+				Namespace: metav1.NamespaceSystem,
+			},
+		},
+	}
+
+	ciliumBaseName   = "cilium-etcd-client"
+	ciliumCertConfig = certutil.Config{
+		CommonName: "cilium-etcd-client",
+		Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+)
+
+func init() {
+	// self-register in the CNI plugins registry
+	cni.Registry.Register("cilium", EnsureCiliumAddon)
+}
+
+// EnsureCiliumAddon creates the cilium addons
+func EnsureCiliumAddon(cfg *config.KubicInitConfiguration, client clientset.Interface) error {
+	etcdDir := strings.Join([]string{cfg.Certificates.Directory, "/etcd"}, "")
+	var ciliumConfigMapBytes, ciliumDaemonSetBytes []byte
+
+	caCert, caKey, err := pkiutil.TryLoadCertAndKeyFromDisk(etcdDir, "ca")
+	if err != nil {
+		fmt.Printf("etcd generation retrieval failed %s", err)
+	}
+
+	cert, key, err := pkiutil.NewCertAndKey(caCert, caKey, &ciliumCertConfig)
+	if err != nil {
+		fmt.Printf("error when creating etcd client certificate for cilium %s", err)
+	}
+
+	pkiutil.WriteCertAndKey(etcdDir, ciliumBaseName, cert, key)
+	if err := createServiceAccount(client); err != nil {
+		return fmt.Errorf("error when creating cilium service account: %v", err)
+	}
+
+	etcdServer, err := cfg.GetBindIP()
+	if err != nil {
+		return fmt.Errorf("error when trying to retrieve etcd server IP %s", err)
+	}
+
+	ciliumConfigMapBytes, err = kubeadmutil.ParseTemplate(CiliumConfigMap,
+		struct {
+			EtcdServer string
+		}{
+			etcdServer.String(),
+		})
+
+	if err != nil {
+		return fmt.Errorf("error when parsing cilium configmap template: %v", err)
+	}
+
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cilium-secret",
```

Comment:
Just a niptick: I would refactor this into a constant:

```suggestion
			Name:      CiliumCertSecret,
```

and then you could define

```Go
const (
    # ...
    CiliumCertSecret = "cilium-secret"
)
```

and then you could use the same constant in the template with

```Go
ciliumDaemonSetBytes, err = kubeadmutil.ParseTemplate(CiliumDaemonSet,
		struct {
			Image      string
			ConfDir    string
			BinDir     string
                        SecretName string
		}{
			cfg.Network.Cni.Image,
			cfg.Network.Cni.ConfDir,
			cfg.Network.Cni.BinDir,
                        CiliumCertSecret,
		})
```

so you could use `{ .SecretName }` in the template.